add post about compression (#46)
dandi/dandi-about: _posts/2022-03-20-compression.md

Assessing NWB compression options
=================================

When submitting large dandisets, it is important that users consider data compression options, which can substantially 
reduce the size of the files. Making files smaller reduces the burden of the DANDI Archive, and make the files more 
convenient to download. HDF5 has the ability to apply chunking and compression to datasets, which breaks large datasets
into smaller chunks and applies lossless compression to each chunk. This approach reduces the volume of the HDF5 
file without changing the values of the dataset, and in some cases can even improve read or write speeds above naive
uncompressed binary data. HDF5 ships with the compression algorithms GZIP and LZ4, and can support additional

> dynamically loaded filters. DANDI has requested that large datasets be compressed, as it reduces the main These 

Author: ```suggestion
dynamically loaded filters. DANDI has requested that large datasets be compressed. These 
```

By how much can the volume of my data be reduced?

What options are available and how can I access these options?

What compression algorithms and settings are best for me?

The answers to all of these questions depend on the specific use-cases, but the analysis below should help. 

There are generally 3 metrics you might care about: file size, read speed, and write speed. You also might care how
accessible your dataset is, i.e. whether the HDF5 library would be able to read your file out of the box or require
the installation of a dynamic filter. There are several parameters we can explore. Let's start with compression
algorithm (gzip, lz4, zstd, etc.) and level of algorithm. Increasing the level attempts to make the file smaller at the
cost of additional compute time.

## Approach

We ran a scan over several popular compression algorithms using the
[h5plugin](http://www.silx.org/doc/hdf5plugin/latest/) library, which automatically compiles 
and installs several algorithms and makes them available to [h5py](https://docs.h5py.org/en/stable/index.html). If using
Python, we highly recommend using this package, as it vastly simplifies the installation process for a good variety of

> popular compression algorithms.For test data, we are using ap data from a Neuropixel probe recorded using the SpikeGLX

Author: ```suggestion
popular compression algorithms. For test data, we are using ap data from a Neuropixel probe recorded using the SpikeGLX
```

acquisition system from [dandiset 000053](https://dandiarchive.org/dandiset/000053/0.210819.0345), which is high-pass
voltage data prior to spike-sorting. We expect this to be a common use-case for NWB. Here, we are relying on h5py to
automatically determine chunk shape, and shuffle is off.

## Results

<img src="media/eval_compressions.png"/>
Figure 1. The black circle shows data that is chunked but not compressed. Each color represents a different 
algorithm and the lightness/darness of the color indicates compression level.

Of the filters we tested, zstd shows the best overall results. zstd is about tied with
gzip and blosz zlib for the best data compression ratio (about 45%). At low to moderate levels of compression, zstd 

> is also close to optimal for read and write speed. For higher levels, zstd write speed increases dramatically without 

Author: ```suggestion
is also close to the best for read and write speed. For higher levels, zstd write speed increases dramatically without 
```

much additional space savings, so our recommendations for best all-around performance would be zstd level 4. zstd is 
faster than naive binary read, but if you are optimizing for read speed above all else, blosc lz4 is the only 
algorithm we tested that can be read substantially faster than zstd (1.25x). If using blosz lz4, level 4 appears to 
offer a good trade-off between size and speed. Note that the default gzip performs substantially worse in read time 
(about 1.8x zstd), but at moderate levels (e.g. the default 4) has about the same performance in file size and write 
time, making it an adequate alternative for users unable to configure custom dynamic filters or anticipating that the 

> readers of their data may not be able to do so.

Author: ```suggestion
readers of their data may not be able to do so. blos zstd gets a slightly better compression ratio than zstd (without blosc), resulting in smaller files, but blos zstd takes much longer to write the data, particularly at higher compression levels.
```

> ## Discussion
> 
> There are a variety of compression options supported by HDF5. zstd level 4 will probably be the best for most NWB 
> users, there is no one option that is best in every category and every option comes with trade-offs.
> 
> ### MATLAB users
> MATLAB has officially supported reading HDF5 datasets with dynamically loaded filters since 2015a, however this 
> requires some custom configuration and we have not yet found a convenient way to do this. We have heard that MATLAB 
> 2022a will be the first release to support _writing_ HDF5 Datasets with dynamically loaded filters. More tutorials 
> to come once we figure that out.

Author: ```suggestion
There are a variety of compression options supported by HDF5. zstd level 4 will probably be the best for most NWB 
users, there is no one option that is best in every category and every option comes with trade-offs.
```
